metadataserver/cni/env.go

```diff
@@ -0,0 +1,249 @@
+	var k8sArgs utils.K8sArgs
+	err := types.LoadArgs(args.Args, &k8sArgs)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to parse CNI k8s args")
+	}
+	logrus.Debugf("k8s args %#v", k8sArgs)
+
+	ctx := context.Background()
+	defer ctx.Done()
+
+	return utils.GetPod(ctx, kubeletAPIPodsURL, k8sArgs)
+}
+
+func generateTokenKeySaltIfRequired(pod *v1.Pod) (string, error) {
+	fname := path.Join(tt.TitusEnvironmentsDir, fmt.Sprintf("%s.salt", pod.Name))
+
+	_, err := os.Stat(fname)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return "", errors.Wrap(err, "Unable to stat "+fname)
+		}
+
+		fd, err := os.OpenFile(fname, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644) // nolint: gosec
+		if err != nil {
+			return "", err
+		}
+
+		if _, err = fmt.Fprint(fd, uuid.New().String()); err != nil {
+			return "", err
+		}
+
+		if err = fd.Close(); err != nil {
+			return "", err
+		}
+	}
+
+	data, err := ioutil.ReadFile(fname)
+	return string(data), err
+}
+
+func extractEnv(prev *current.Result, pod *v1.Pod) (map[string]string, error) {
+	env := map[string]string{}
+
+	env["LISTEN_PORT"] = "80"
+
+	env["PEER_NAMESPACE"] = getPeerNsPath(pod.Name)
+	env["EC2_REGION"] = os.Getenv("EC2_REGION")
+
+	var ok bool
+
+	for _, ip := range prev.IPs {
+		switch ip.Version {
+		case "4":
+			env["EC2_LOCAL_IPV4"] = ip.Address.String()
+			ok = true
+		case "6":
+			env["EC2_IPV6S"] = ip.Address.String()
+			ok = true
+		}
+	}
+
+	if !ok {
+		return nil, errors.New("No IP information in chained CNI result")
+	}
+
+	// pod specific configuration
+	env["TITUS_TASK_INSTANCE_ID"] = pod.Name
+
+	// FIXME(manas) make this an annotation in the admission webhook
+	env["REQUIRE_TOKEN"] = pod.Annotations[mt.RequireTokenAnnotation]
+	salt, err := generateTokenKeySaltIfRequired(pod)
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to generate required token salt")
+	}
+	env["TOKEN_KEY_SALT"] = salt
+
+	env["TITUS_IAM_ROLE"] = pod.Annotations[mt.IamRoleArnAnnotation]
+	env["X_FORWARDED_FOR_BLOCKING_MODE"] = pod.Annotations[mt.XForwardedForBlockingModeAnnotation]
+
+	// assume we are on a metatron enabled host if we find /metatron
+	_, err = os.Stat("/metatron")
+	if os.IsNotExist(err) {
+		env[mt.TitusMetatronVariableName] = "false"
+	}
+	env[mt.TitusMetatronVariableName] = "true"
+
+	return env, nil
+}
+
+func setupEnv(pod *v1.Pod, env map[string]string) error {
+	err := writeEnvFile(pod, env)
+	if err != nil {
+		return errors.Wrap(err, "Unable to write environment file")
+	}
+
+	metatronEnv := env[mt.TitusMetatronVariableName]
+
+	if len(metatronEnv) == 0 {
+		return nil
+	}
+
+	metatronOn, err := strconv.ParseBool(metatronEnv)
+	if err != nil {
+		return errors.Wrap(err, "Unable to tell if metatron is enabled")
+	}
+
+	if !metatronOn {
+		return nil
+	}
+
+	err = writeContainerInfo(pod)
+	if err != nil {
+		return errors.Wrap(err, "Unable to write containerInfo file")
+	}
+
+	return nil
+}
+
+func writeEnvFile(pod *v1.Pod, env map[string]string) error {
```

Comment:
as I read more and more of this, I am beginning to think that the place where this should be should instead be in a CRI-wrapper. We should create (and delete) state as the pod sandbox is created.

Reply:
(as well as the state file).

Reply:
We cannot create this file till we have allocated an ENI for the pod. My plan, now that I also looking at setting up netflix environment, was to have this CNI write the file out, and use it to set environment for containers within the pod.

If we use a CRI-wrapper, we will have to make the VPC service calls from the wrapper.

Reply:
If we have the CRI write out the environment variable file after the sandbox is created, we can do this. It's more work, but it might be the "better" thing to do.

Then I think we have two sidecars that we add:
1. The (unprivileged) IMDS metadata service that's network namespace is set to host network namespace (but it the FD passed to it is a shared volume, like `/run/imds.sock`)
2. A privileged sidecar (CAP_NET_ADMIN), that runs inside of the container and does unshare and just proxies to `/run/imds.sock`. I know you suggested we do something like this earlier, and you were right. I didn't consider a CRI "interposer" at the time. Without the CRI interposer you cannot do #1, as you cannot have some containers in the pod have the host namespace, and some have the other namespace. 

The "only" thing you need to do it protect the proxy from having its /proc/self/ns/net stolen by another process in the pod. In order to do that, we need to do something like set dumpable to false (https://github.com/containerd/cri/issues/1453).

In addition to that, you need to get "ordering" right, and ensure that #1 and #2 run before user code. 

----
Okay, this is quickly becoming a nightmare of an implementation story. 
----
Can we just intercept RunSandbox, write the environment file, and then start the systemd unit in the interposer?

Reply:
Does this need to be done via a CRI, or could the CNI launch containers to do this?  They wouldn't be visible to k8s, or to the pod, but that's a feature here, not a bug, I think.

I think the sidecar / socket approach is probably the most secure way of doing this, but doing this via a CRI wrapper seems like overkill.